crates/subspace-runtime/src/lib.rs

```diff
@@ -746,6 +747,27 @@ fn extract_bundles(extrinsics: Vec<OpaqueExtrinsic>) -> Vec<OpaqueBundle> {
         .collect()
 }
 
+fn extract_pre_digest<Block: BlockT>(
+    header: Block::Header,
+) -> Result<PreDigest<FarmerPublicKey>, PreDigestError> {
+    // genesis block doesn't contain a pre digest so let's generate a
+    // dummy one to not break any invariants in the rest of the code
+    if header.number().is_zero() {
+        return Ok(PreDigest::genesis_pre_digest());
+    }
+
+    let mut pre_digest: Option<_> = None;
+    for log in header.digest().logs() {
+        match (log.as_subspace_pre_digest(), pre_digest.is_some()) {
+            (Some(_), true) => return Err(PreDigestError::MultipleDigests),
+            (None, _) => {}
+            (s, false) => pre_digest = s,
+        }
+    }
+
+    pre_digest.ok_or(PreDigestError::NoDigest)
+}
```

Comment:
I'd return randomness from here and panic if pre-digest is not found, it is really a violation of the protocol design.

Reply:
Resolved in https://github.com/subspace/subspace/pull/218/commits/5552d30ffef5897136c9ec023afcaa7e6725115b